core/dbt/adapters/base/impl.py

```diff
@@ -997,25 +998,52 @@ def _catalog_filter_table(
         """
         return table.where(_catalog_filter_schemas(manifest))
 
-    def _get_catalog_information_schemas(
-        self, manifest: Manifest
-    ) -> List[InformationSchema]:
-        return list(self._get_cache_schemas(manifest).keys())
+    def _get_one_catalog(
+        self,
+        information_schema: InformationSchema,
+        schemas: Set[str],
+        manifest: Manifest,
+    ) -> agate.Table:
 
-    def get_catalog(self, manifest: Manifest) -> agate.Table:
-        """Get the catalog for this manifest by running the get catalog macro.
-        Returns an agate.Table of catalog information.
-        """
-        information_schemas = self._get_catalog_information_schemas(manifest)
-        # make it a list so macros can index into it.
-        kwargs = {'information_schemas': information_schemas}
-        table = self.execute_macro(GET_CATALOG_MACRO_NAME,
-                                   kwargs=kwargs,
-                                   release=True)
+        name = '.'.join([
+            str(information_schema.database),
+            'information_schema'
+        ])
+
+        # calculate the possible schemas for a given schema name
+        all_schema_names: Set[str] = set()
+        for schema in schemas:
+            all_schema_names.update({schema, schema.lower(), schema.upper()})
+
+        with self.connection_named(name):
+            kwargs = {
+                'information_schema': information_schema,
+                'schemas': all_schema_names
+            }
+            table = self.execute_macro(GET_CATALOG_MACRO_NAME,
+                                       kwargs=kwargs,
+                                       release=True)
 
         results = self._catalog_filter_table(table, manifest)
         return results
 
+    def get_catalog(self, manifest: Manifest) -> agate.Table:
+        # snowflake is super slow. split it out into the specified threads
+        num_threads = self.config.threads
+        schema_map = self._get_cache_schemas(manifest)
+        catalogs: agate.Table = agate.Table(rows=[])
+
+        with ThreadPoolExecutor(max_workers=num_threads) as executor:
+            futures = [
+                executor.submit(self._get_one_catalog, info, schemas, manifest)
+                for info, schemas in schema_map.items() if len(schemas) > 0
+            ]
+            for future in as_completed(futures):
```

Comment:
what would happen here if one of these futures fails (eg. with a database error)?

Reply:
Well, this is my expectation:
- future.result() will raise the DatabaseException
- the exception will go through get_catalog, etc
- eventually dbt will hit the context manager that cancels connections
- ultimately dbt exits with an error about the DatabaseException

If we wanted, we could change it to check if the result was an exception, and if so store it until the end so all the threads are done. Then we could just raise the first one, or something.

Reply:
> If we wanted, we could change it to check if the result was an exception, and if so store it until the end so all the threads are done. Then we could just raise the first one, or something.

Yeah - I like this idea! I was initially worried that an exception here could cause dbt to hang, but that's definitely not the case. I was getting mixed up with how we do exception handling in the multithreaded job runner.

I think it would be great if the `docs generate` command wrote out a catalog.json file if _any_ of the catalog queries succeed. This has proven to be a gnarly issue on BigQuery in particular, in which obscure additional permissions are required to fetch data from  information schema tables.

Check out some relevant threads from dbt Slack:

![Screen Shot 2020-01-31 at 9 40 56 AM](https://user-images.githubusercontent.com/1541139/73547918-e37b6580-440d-11ea-820f-c795ed58398e.png)
![Screen Shot 2020-01-31 at 9 41 09 AM](https://user-images.githubusercontent.com/1541139/73547919-e37b6580-440d-11ea-9b7f-8664f7725bc8.png)
![Screen Shot 2020-01-31 at 9 41 19 AM](https://user-images.githubusercontent.com/1541139/73547920-e37b6580-440d-11ea-8787-b876f539cc42.png)

I'm picturing something where we catch DatabaseExceptions raised from these futures and log an error message out, but continue waiting for the other futures to return. If any of them return successfully, then we can proceed on to writing out the catalog from the task. After the catalog is written, the task should exit with a nonzero exit code.

Do you buy that?

Reply:
>I'm picturing something where we catch DatabaseExceptions raised from these futures and log an error message out, but continue waiting for the other futures to return. If any of them return successfully, then we can proceed on to writing out the catalog from the task. After the catalog is written, the task should exit with a nonzero exit code.

Sure, we can do that. We should probably make it a warning I guess, so people can opt-in to it being fatal with `--strict`.

Reply:
Yeah, good idea!